File: ChromaSchemes/tooling/previewModelBuilder.ts
/**
 * PreviewModelBuilder – builds preview-ready model payloads.
 *
 * Description:
 *   Converts colour-set and role-binding data into the preview model consumed
 *   by the UI renderer.
 *
 * Inputs:
 *   - build(colourSet, roles)
 *     * colourSet: normalised colour entries.
 *     * roles: resolved role-binding map.
 *
 * Outputs:
 *   A preview model object with pane definitions and colour references.
 *
 * Example:
 *   const builder = new PreviewModelBuilder();
 *   const preview = builder.build(colourSet, roles);
 */
import { IPreviewModelBuilder } from './interfaces';

export class PreviewModelBuilder implements IPreviewModelBuilder {
  build(colourSet: Array<Record<string, unknown>>, roles: Record<string, unknown>) {
    const borderRef =
      (roles.border_fg as { layers?: { fg?: { colourRef?: string } } } | undefined)?.layers?.fg?.colourRef ?? 'c0';

    return {
      model: 'MultiPanePreviewModel',
      panes: Array.from({ length: 4 }, (_, idx) => ({
        paneId: `pane_${idx}`,
        layout: 'fractal_quadrant',
        border: { colourRef: borderRef },
        fill: { colourRef: String(colourSet[idx % colourSet.length]?.id ?? 'c0') },
      })),
    };
  }
}

File: ChromaSchemes/core/preview_model_builder.py
"""Build deterministic multi-pane preview models."""

from __future__ import annotations


class PreviewModelBuilder:
    """Generate a pure multi-pane fractal quadrant preview model."""

    def build(self, colour_set: list[dict], roles: dict) -> dict:
        refs = [entry["id"] for entry in colour_set]
        border_ref = roles.get("border_fg", {}).get("layers", {}).get("fg", {}).get("colourRef", refs[0])

        panes = []
        for idx in range(4):
            panes.append(
                {
                    "paneId": f"pane_{idx}",
                    "layout": "fractal_quadrant",
                    "roleUsageHints": {
                        "backgroundRole": ["primary_surface", "secondary_surface", "tertiary_surface", "accent_surface"][idx],
                        "interactiveRole": "interactive_bg",
                        "textRole": "menu_fg",
                    },
                    "border": {"colourRef": border_ref},
                    "fill": {"colourRef": refs[idx % len(refs)]},
                }
            )

        return {"model": "MultiPanePreviewModel", "panes": panes}
